### ISSUE TITLE
fix: Granite 4 Vision text-only forward intermittently produces non-finite logits

### ISSUE THREAD

**Issue by inureyes** [lablup/mlxcel#1211]
## Problem / Background

`text_only_forward_produces_finite_logits` in `tests/granite4_vision_parity.rs:76-104` fails intermittently against the real `granite-4.0-3b-vision-4bit` checkpoint. The assertion is `tests/granite4_vision_parity.rs:100`, message `text-only logits must be finite`. A single load plus a single prompt forward of a shipped model family produces a non-finite maximum logit roughly one run in eight, which is a correctness signal about the model path rather than a cosmetic test problem.

The test self-skips when `models/granite-4.0-3b-vision-4bit` is absent: `model_dir()` at `tests/granite4_vision_parity.rs:30-38` resolves through `common::repo_model_dir` (`tests/common/mod.rs:5-21`), `eprintln!`s, and returns `None`, and the test returns early. The nightly `nightly-verify.yml` job therefore never executes it, because the self-hosted runner carries no model weights, exactly as `.github/workflows/nightly-verify.yml:55-57` documents. The failure only surfaces on a developer machine running `make verify` (or `make verify-test`) with the checkpoint present locally.

## Current Behavior

What the test does, in order: `mlxcel::load_model(&dir)`, tokenize `"Hello, world."` with BOS, build a `[1, len]` i32 array, `LanguageModel::make_caches`, one `LanguageModel::forward` over the whole prompt, `eval`, slice the last position's logits row, `max_all`, `eval`, then assert `item_f32(&max).is_finite()`.

The path the failing call actually takes contains no vision code:

- `src/loaded_model.rs:90` holds the `LoadedModel::Granite4VisionVLM(vision::Granite4VisionVLModel)` variant, and the dispatch macro at `src/loaded_model.rs:259` forwards `LanguageModel` methods to the inner model.
- `src/vision/granite4_vision.rs:263-271` implements `LanguageModel::forward` as a straight delegation to `self.text_model.forward(input_ids, caches, mask)`. The DeepStack / window-QFormer injection lives in `inject_forward`, reached only through `forward_with_embeddings` (`src/vision/granite4_vision.rs:273-281`), which this test never calls.
- The text backbone is `granitemoehybrid`, an interleaved Mamba2 (SSM) plus GQA attention hybrid (`src/models/granitemoehybrid.rs:15-26`). Its `LanguageModel::forward` at `src/models/granitemoehybrid.rs:1514-1526` ignores the external `KVCache` slice entirely and runs through `sequence_state.with_sequence_state(None, ...)` into `forward_with_caches` (`src/models/granitemoehybrid.rs:1053-1062`). `make_caches` at `src/models/granitemoehybrid.rs:1528-1534` replaces the model-internal mixed cache via `GraniteMoeHybridModel::make_caches` (`src/models/granitemoehybrid.rs:1049`) and hands back empty `KVCache`s. The mixed per-layer cache holds `Mamba2Cache` recurrent conv/SSM state for mamba layers and `KVCache` for attention layers (`src/models/granitemoehybrid.rs:66-68`).
- Loading runs through `load_granite4_vision_vlm` at `src/loading/vlm_granite4_vision.rs:105`, which inherits the top-level `quantization` block into `text_config` when the latter lacks one (`src/loading/vlm_granite4_vision.rs:113-118`), parses the text config as `models::granitemoehybrid::ModelArgs` (`src/loading/vlm_granite4_vision.rs:119`), and loads weights at `src/loading/vlm_granite4_vision.rs:134`. The checkpoint under test is 4-bit quantized.

## Measured Evidence

Measured 2026-08-18 on an Apple M5 Max (`Mac17,7`, 18 logical CPUs, macOS 26.6.1) at commit `5dfcb390` on `main`, by running the compiled test binary directly under `[profile.test-fast]` with `--features metal,accelerate`, 40 runs per arm, on an otherwise idle machine:

| libtest threads | runs | failures | rate |
|---|---|---|---|
| default (18) | 40 | 6 | 15% |
| `--test-threads=1` | 40 | 4 | 10% |

The point of that table is the comparison, not the individual rates: the flake does not go away when libtest is serialized. Both arms fail at the same order of magnitude, roughly one run in eight across the combined 10 of 80.

## Relationship to Existing Issues

**This is not the concurrency problem tracked in #1092, and PR #1210 does not fix it.** #1210 adds `--test-threads=1` to `verify-test` because the macOS gate's `mlxcel-core` binary died with `SIGSEGV` under 18 concurrent libtest workers. The evidence table above was collected specifically to separate the two: with libtest fully serialized, exactly the configuration #1210 lands, this assertion still fails 4 times in 40. Do not close this as fixed when #1210 merges, and do not attribute a red `verify-test` on this test to #1092.

**#997 is the pre-existing broader tracker** (`fix: Resolve flaky text_only_forward_produces_finite_logits in granite4_vision and hunyuan_vl parity tests`, open, `priority:medium`). It covers this same assertion plus the parallel `tests/hunyuan_vl_parity.rs:105` case, and already records a 4-in-20 isolated failure rate for granite4 measured 2026-08-03 on `e909998f`. This issue narrows the granite4 half to a single-family scope with a thread-count-controlled measurement at the current commit. A maintainer should consolidate: either close this into #997 after copying the table across, or split #997 down to the hunyuan case. Do not have two people bisect the same forward pass.

## Open Questions

These are unknown as of filing and should be answered by the investigation rather than assumed:

- **NaN or Inf?** The assertion only checks `is_finite()`, so the failing value's identity was never captured. `NaN` points at a `0 * Inf`, `Inf - Inf`, or an invalid `exp`/`rsqrt` input; a raw `Inf` points at an overflow or a division by a zero denominator. These have different suspects and the distinction should be the first thing established.
- **Load path or forward path?** Whether the non-finite value originates in weight loading and dequantization (`src/loading/vlm_granite4_vision.rs:105-254`, the 4-bit path) or is produced during the forward pass itself (`src/models/granitemoehybrid.rs:1053`, including the Mamba2 conv/SSM state initialization). A load-side defect would show as bad weights on a freshly loaded model even before any forward runs; a forward-side defect would show clean weights and a non-finite intermediate at some layer.
- **Which other entry points?** Whether the full VLM path with real image input (`forward_with_embeddings` into `inject_forward`, `src/vision/granite4_vision.rs:273-281`), and whether longer prompts than `"Hello, world."`, show the same rate, a different rate, or none.

## Scope

**In scope:** determining the root cause of the non-finite logit for `granite4_vision` on the text-only forward, and fixing it if it is a runtime defect, across `src/vision/granite4_vision.rs`, `src/models/granitemoehybrid.rs`, and `src/loading/vlm_granite4_vision.rs`.

**Out of scope:** the `tests/hunyuan_vl_parity.rs:105` counterpart, which stays with #997 unless the investigation proves a shared root cause. The libtest serialization change in PR #1210, which is a separate fix for a separate failure mode.

## Investigation Notes

- **Do not "fix" this by loosening the assertion.** Replacing `is_finite()` with a tolerance, marking the test `#[ignore]`, or gating it behind an env var hides a real-model correctness signal. If the test is changed at all, it should be changed to capture more information on failure, not less.
- **Capture the value first.** Changing the assertion message to print the actual `f32` (and ideally the count of non-finite entries in the full logits row) costs nothing and answers the NaN-versus-Inf question on the next reproduction.
- **Reuse the existing model-dir helper.** Any new test or probe should go through `common::repo_model_dir` (`tests/common/mod.rs:5`) so it self-skips consistently, rather than hardcoding a path.
- **The failure is not thread-count sensitive**, per the table above, so time spent on libtest concurrency, `--jobs`, or cross-target contention is time wasted. #997's original cross-target contention hypothesis was already ruled out for granite4 by its 2026-08-03 isolated measurement.
- **Nondeterminism inside one process on one checkpoint** means the input is fixed and the weights are fixed, so the varying quantity is runtime state: allocator contents, Metal command buffer scheduling, uninitialized memory, or recurrent state that is not fully zeroed at construction. The Mamba2 conv/SSM cache is the one piece of per-run state in this path that a fresh `KVCache` is not.

## Acceptance Criteria

- [ ] The failing value is characterized: the issue thread records whether it is `NaN` or `+/-Inf`, and how many entries of the last-position logits row are non-finite when it fires.
- [ ] The origin is localized to either the weight-loading/dequantization path or the forward pass, with the specific layer index and operation named.
- [ ] Root cause identified and stated, distinguishing a numerical defect in the granite4_vision / granitemoehybrid path from a runtime or harness artifact.
- [ ] If it is a runtime defect, it is fixed in the shipping code path (not only in the test), so that CLI generation through `mlxcel generate -m models/granite-4.0-3b-vision-4bit` benefits, and a regression test pins the fix.
- [ ] 100 consecutive runs of `text_only_forward_produces_finite_logits` pass on an Apple Silicon host with the checkpoint present, at both default and `--test-threads=1`.
- [ ] The relationship to #997 is resolved: one of the two issues is closed with a pointer to the other, and the hunyuan_vl case has an explicit verdict (same root cause or not).
- [ ] `make verify` is green across two consecutive runs on a machine that has the checkpoint.

## Verification

```bash
# Fetch the checkpoint if absent (the test silently self-skips without it).
./target/release/mlxcel download mlx-community/granite-4.0-3b-vision-4bit

# Single reproduction. Passes most of the time; repeat it.
cargo test --profile test-fast --features metal,accelerate -p mlxcel --test granite4_vision_parity

# Rate measurement. Build once, then run the compiled binary repeatedly so the
# sample measures the test rather than cargo. Both arms should be run.
cargo test --profile test-fast --features metal,accelerate -p mlxcel \
  --test granite4_vision_parity --no-run
BIN=$(ls -t target/test-fast/deps/granite4_vision_parity-* | grep -v '\.d$' | head -1)
FAILED=0
for i in $(seq 1 40); do
  "$BIN" --exact text_only_forward_produces_finite_logits >/dev/null 2>&1 || FAILED=$((FAILED+1))
done
echo "default threads: $FAILED / 40 failed"

FAILED=0
for i in $(seq 1 40); do
  "$BIN" --test-threads=1 --exact text_only_forward_produces_finite_logits >/dev/null 2>&1 || FAILED=$((FAILED+1))
done
echo "single thread: $FAILED / 40 failed"
```

A pass is 0 failures in 40 for both arms after the fix. The baseline recorded above is 6/40 and 4/40 at `5dfcb390`. Note that a run on a host without `models/granite-4.0-3b-vision-4bit` will report 0/40 vacuously, because the test skips, so confirm the checkpoint directory exists before trusting a green result.

## Technical Considerations

- Environment for the recorded measurement: Apple M5 Max (`Mac17,7`), 18 logical CPUs, macOS 26.6.1, commit `5dfcb390` on `main`, `[profile.test-fast]`, `--features metal,accelerate`, machine otherwise idle.
- Granite 4 Vision was ported in #541. The `granitemoehybrid` text backbone is shared with the non-VLM Granite 4 checkpoints (`granite-4.0-h-tiny-4bit`, `granite-4.1-3b-4bit`, `granite-4.1-8b-4bit`), so a defect in the text path would not be confined to the VLM. Checking whether a text-only Granite 4 checkpoint shows the same behavior is a cheap way to widen or narrow the blast radius.
- Related: #997 (broader tracker, includes hunyuan_vl), #1092 (nightly red), #1210 (libtest serialization, explicitly not a fix for this).

**Comment by inureyes** [comment]
Duplicate of #997, which has tracked this assertion since 2026-08-02 and also covers the `hunyuan_vl_parity.rs:105` counterpart. Filed this before finding it.

The one thing here that #997 did not have, the thread-count-controlled 40-run-per-arm table at `5dfcb390` and the path reading that rules out the vision code, is now a comment on #997. Continue there.
